src/core/raster/qgsrasteriterator.cpp

```diff
@@ -169,11 +181,19 @@ bool QgsRasterIterator::readNextRasterPartInternal( int bandNumber, int &nCols,
   }
 
   //read data block
-
   tileTopLeftColumn = pInfo.currentCol;
   tileTopLeftRow = pInfo.currentRow;
+
   tileColumns = static_cast< int >( std::min( static_cast< qgssize >( mMaximumTileWidth ), pInfo.nCols - tileTopLeftColumn ) );
   tileRows = static_cast< int >( std::min( static_cast< qgssize >( mMaximumTileHeight ), pInfo.nRows - tileTopLeftRow ) );
+
+  if ( mResamplingFactor > 1 )
+  {
+    // Round down tile dimensions to resampling factor
+    tileColumns = ( tileColumns / mResamplingFactor ) * mResamplingFactor;
```

Comment:
This approach won't work if mResamplingFactor > mMaximumTileWidth or mMaximumTileHeight. It could be problematic in particular for scanline access patterns where mMaximumTileHeight is set to 1.

Reply:
how would you suggest we handle that? Limit the resampling factor to at most the tile size? Or just add a documentation warning and require the caller to check this?

Reply:
I think you should rather keep the tile dimension but increase the queried extent

Reply:
>  increase the queried extent

what if that falls outside the raster bounds?

Reply:
> what if that falls outside the raster bounds?

clamp and modify the shape of the returned block ?

Let's say you have a raster of 5x5, you set resamplingFactor=2, mMaximumTileWidth=2 and mMaximumTileHeight=2, then I'd expect to have:
- one block of size 2x2 covering (0,0) included to (4,4) excluded
- one block of size 1x2 covering (4,0) included to (5,4) excluded
- one block of size 2x1 covering (0,4) included to (4,5) excluded
- one block of size 1x1 covering (4,4) included to (5,5) excluded

By the way reviewing the code, there's a copy and paste issue at https://github.com/qgis/QGIS/blob/c244bc185d3de346ef2ad96eb30eeea824d934c5/src/core/raster/qgsrasteriterator.h#L215 . blockCountHeight() should return blockCountHeight not width.

Reply:
@rouault 

I **think** I disagree :smile: 

I've just renamed the methods from "resampling" to "snap to pixel factor" so that it's more clear what these are intended to do. Because then when we look at your example:

> one block of size 2x2 covering (0,0) included to (4,4) excluded
one block of size 1x2 covering (4,0) included to (5,4) excluded
one block of size 2x1 covering (0,4) included to (4,5) excluded
one block of size 1x1 covering (4,4) included to (5,5) excluded

It's hopefully(?!) clear that ONLY the 2x2 block should be returned here, as the others will snap down to nothing.

What do you think?

> there's a copy and paste issue at

Thanks, fixed

Reply:
oooooh, it seems I totally misunderstood what this was doing. I assumed it returned a sampled down version of the raster. I now get it. Can we find an alternative name to "resampling factor" to waive any ambiguity ? "snapping factor" maybe?

Reply:
@rouault that's what I did in https://github.com/qgis/QGIS/pull/59733/commits/c7fffb2dd9d3ec2803bca0f2ba926e742ba7da42

Reply:
fine, LGTM then